app/ante/cosmos/fees.go

```diff
@@ -0,0 +1,235 @@
+		} else if !feeGranter.Equals(feePayer) {
+			err := dfd.feegrantKeeper.UseGrantedFees(ctx, feeGranter, feePayer, fees, sdkTx.GetMsgs())
+			if err != nil {
+				return errorsmod.Wrapf(err, "%s does not not allow to pay fees for %s", feeGranter, feePayer)
+			}
+		}
+
+		deductFeesFrom = feeGranter
+	}
+
+	deductFeesFromAcc := dfd.accountKeeper.GetAccount(ctx, deductFeesFrom)
+	if deductFeesFromAcc == nil {
+		return errortypes.ErrUnknownAddress.Wrapf("fee payer address: %s does not exist", deductFeesFrom)
+	}
+
+	// deduct the fees
+	if !fees.IsZero() {
+		if err := dfd.deductFeesFromBalanceOrUnclaimedRewards(ctx, deductFeesFromAcc, fees); err != nil {
+			return err
+		}
+	}
+
+	events := sdk.Events{
+		sdk.NewEvent(
+			sdk.EventTypeTx,
+			sdk.NewAttribute(sdk.AttributeKeyFee, fees.String()),
+			sdk.NewAttribute(sdk.AttributeKeyFeePayer, deductFeesFrom.String()),
+		),
+	}
+	ctx.EventManager().EmitEvents(events)
+
+	return nil
+}
+
+// deductFeesFromBalanceOrUnclaimedRewards tries to deduct the fees from the account balance.
+// If the account balance is not enough, it tries to claim enough staking rewards to cover the fees.
+// If the account does not have sufficient staking rewards, it returns an error.
+func (dfd DeductFeeDecorator) deductFeesFromBalanceOrUnclaimedRewards(ctx sdk.Context, deductFeesFromAcc authtypes.AccountI, fees sdk.Coins) error {
+	balance := dfd.bankKeeper.GetBalance(ctx, deductFeesFromAcc.GetAddress(), fees[0].Denom)
+	if balance.IsLT(fees[0]) {
+		difference := fees[0].Sub(balance)
+		// Try to claim enough staking rewards to cover the difference between the
+		// transaction cost and the account balance.
+		err := evm.ClaimSufficientStakingRewards(ctx, dfd.stakingKeeper, dfd.distributionKeeper, deductFeesFromAcc.GetAddress(), sdk.Coins{difference})
+		if err != nil {
+			return errortypes.ErrInsufficientFunds.Wrapf(
+				"insufficient funds and failed to claim sufficient staking rewards for %s to pay for fees; %s",
+				deductFeesFromAcc.GetAddress(),
+				err.Error(),
+			)
+		}
+	}
+	// deduct the fees if possible
+	return authante.DeductFees(dfd.bankKeeper, ctx, deductFeesFromAcc, fees)
+}
+
+// checkTxFeeWithValidatorMinGasPrices implements the default fee logic, where the minimum price per
+// unit of gas is fixed and set by each validator, and the tx priority is computed from the gas price.
+func checkTxFeeWithValidatorMinGasPrices(ctx sdk.Context, tx sdk.Tx) (sdk.Coins, int64, error) {
+	feeTx, ok := tx.(sdk.FeeTx)
+	if !ok {
+		return nil, 0, errorsmod.Wrap(errortypes.ErrTxDecode, "Tx must be a FeeTx")
+	}
+
+	feeCoins := feeTx.GetFee()
+	gas := feeTx.GetGas()
+
+	// Ensure that the provided fees meets a minimum threshold for the validator,
+	// if this is a CheckTx. This is only for local mempool purposes, and thus
+	// is only ran on CheckTx.
+	if ctx.IsCheckTx() {
+		minGasPrices := ctx.MinGasPrices()
+		if !minGasPrices.IsZero() {
```

Comment:
this is hard to read with all the nested `if`, could you refactor by using the inverse?

Reply:
please avoid nested `if`s. You can do something like

```
if !ctx.IsCheckTx() {
return feeCoins, priority, nil
}

// ...
```

Reply:
refactored this into a separate function for better readability